Blog rwd basics (#2873)
GoogleChrome/web.dev: src/site/content/en/blog/responsive-web-design-basics/index.md

---
title: Responsive Web Design Basics
subhead: How to create sites which respond to the needs and capabilities of the device they are viewed on.
description: |
  How to create sites which respond to the needs and capabilities of the device they are viewed on.
date: 2019-02-12
updated: 2020-05-14
authors:
 - petelepage
 - rachelandrew
tags:
 - blog
 - css
 - layout
 - mobile
 - ux
draft: true

> ---
> 
> * [Set the viewport](#viewport)

Reviewer: ```suggestion
---

{# TODO(kayce): Remove this hard-coded ToC once #1983 lands. #}

* [Set the viewport](#viewport)
```
Thank you for hard-coding the ToC. I agree that it's needed and I've expressed to engineering that it's important that we get one across the site and they're working on it.

Reviewer 2: @MichaelSolati is slated to work on it after he finishes the feedback widget.

* [Size content to the viewport](#size-content)
* [Use CSS media queries for responsiveness](#media-queries)
* [How to choose breakpoints](#breakpoints)
* [View media query breakpoints in Chrome DevTools](#devtools)

The use of mobile devices to surf the web continues to grow at an astronomical pace, and these devices are often constrained by display size and require a different approach to how content is laid out on the screen.

Responsive web design, originally defined by [Ethan Marcotte in A List Apart](http://alistapart.com/article/responsive-web-design/), responds to the needs of the users and the devices they're using. The layout changes based on the size and capabilities of the device. For example, on a phone users would see content shown in a single column view; a tablet might show the same content in two columns.

<figure class="w-figure">
  <video controls autoplay loop muted class="w-screenshot">

>     <source src="resize.webm" type="video/webm; codecs=vp8">
>     <source src="resize.mp4" type="video/mp4; codecs=h264">

Reviewer: ```suggestion
    <source src="https://storage.googleapis.com/web-dev-assets/responsive-web-design-basics/resize.mp4" type="video/mp4">
```

* For some reason the webm file is larger than the mp4, whereas usually it's much smaller. Did you convert this or was this already in the google/webfundamentals repo? We can just omit it for now.
* We have a storage bucket where we host videos. We may someday also host images on the bucket (or provide some other GUI).
* Somebody just reported a bug related to the `codecs=vp8` and `codecs=h264` directives so we will no longer be specifying those in our video markup.

Author: These were the files from the existing article, I've not changed them.

## Set the viewport {: #viewport }

Pages optimized for a variety of devices must include a meta viewport tag in the head of the document.
A meta viewport tag gives the browser instructions on how to control the page's dimensions and scaling.

To attempt to provide the best experience, mobile browsers render
the page at a desktop screen width (usually about `980px`, though this varies
across devices), and then try to make the content look better by increasing
font sizes and scaling the content to fit the screen.
This means that font sizes may appear inconsistent to users,
who may have to double-tap or pinch-to-zoom
in order to see and interact with the content.

```html/4
<!DOCTYPE html>
<html lang="en">
  <head>
    …
    <meta name="viewport" content="width=device-width, initial-scale=1">
    …
  </head>
  …

> ```

Reviewer: I forgot to ask, what was your reasoning for why you would have disagreed to providing more context if we didn't have the code highlighting feature? Just want to have a record of it somewhere.

Author: Only that it is then less obvious which bit you are talking about, so if there is no way to highlight it's generally better imo to only show the thing you are talking about, if we can highlight then adding the extra context is indeed preferable.

Using the meta viewport value `width=device-width` instructs the page to match
the screen's width in device-independent pixels. A device (or density) independent pixel being a representation of a single pixel, which may on a high density screen consist of many physical pixels. This allows the page to reflow
content to match different screen sizes, whether rendered on a small mobile
phone or a large desktop monitor.

<figure class="w-figure">

>   <img src="./without-vp-meta.png" width="440" alt="Screenshot of a page with the text hard to read as it is very zoomed out">

Reviewer: This looks grainy on my MacBook Air. Can you re-upload? `1600px` width is safe for ensuring that it has high resolution on all screens (because our `max-width` is `800px`).

![image](https://user-images.githubusercontent.com/4713486/82072900-42977a00-968d-11ea-90e9-920e3cdd891f.png)

Author: these are small images as they are from phone sizes within DevTools. I've redone it however.

  <figcaption class="w-figcaption">
    An example of how the page loads in a device without the viewport meta tag. <a href="https://without-vp-meta.glitch.me/">See this example on Glitch</a>.
  </figcaption>
</figure>

<figure class="w-figure">
  <img src="./with-vp-meta.png" width="440" alt="Screenshot of the same page with the text at a size that can be read">
  <figcaption class="w-figcaption">
    An example of how the page loads in a device with the viewport meta tag. <a href="https://with-vp-meta.glitch.me/">See this example on Glitch</a>.
  </figcaption>
</figure>

[Some browsers](https://css-tricks.com/probably-use-initial-scale1/)
keep the page's width constant when rotating to landscape

> mode, and zoom rather than reflow to fill the screen. Adding the attribute

Reviewer: ```suggestion
mode, and zoom rather than reflow to fill the screen. Adding the value
```
We used `value` to describe `width=device-width` earlier

`initial-scale=1` instructs browsers to establish a 1:1 relationship between CSS
pixels and device-independent pixels regardless of device orientation, and
allows the page to take advantage of the full landscape width.

> Note: To ensure that older browsers can properly parse the attributes,
> use a comma to separate attributes.

Reviewer: ```suggestion
{% Aside 'caution' %}
  To ensure that older browsers can properly parse the attributes,
  use a comma to separate attributes.
{% endAside %}
```
https://web.dev/handbook/web-dev-components/#caution-asides

The [Does not have a `<meta name="viewport">` tag with `width` or `initial-scale`](/viewport/)
Lighthouse audit can help you automate the process of making sure that your HTML documents are using the viewport meta tag correctly.

### Ensure an accessible viewport {: #accessible-viewport }

In addition to setting an `initial-scale`,
you can also set the following attributes on the viewport:

> * `minimum-scale`
> * `maximum-scale`
> * `user-scalable`

Reviewer: ```suggestion
{% Compare 'worse' %}
  * `minimum-scale`
  * `maximum-scale`
  * `user-scalable`
{% endCompare %}
```
This widget will make it clearer that this is an anti-pattern.

When developing a mobile site with a meta viewport tag,
it's easy to accidentally create page content that doesn't quite fit within the specified viewport.
For example, an image that is displayed at a width wider than the viewport
can cause the viewport to scroll horizontally.
You should adjust this content to fit within the width of the viewport,
so that the user does not need to scroll horizontally.

The [Content is not sized correctly for the viewport](/content-width/)
Lighthouse audit can help you automate the process of detecting overflowing content.

### Images {: #images }

An image has fixed dimensions and if it is larger than the viewport will cause a scrollbar.
A common way to deal with this problem is to give all images a `max-width` of `100%`.
This will cause the image to shrink to fit the space it has,
should the viewport size be smaller than the image.
However because the `max-width`, rather than the `width` is `100%`,
the image will not stretch larger than its natural size.
It is generally safe to add the following to your stylesheet
so that you will never have a problem with images causing a scrollbar.

```css
img {

>   max-width: 100%:

Reviewer: ```suggestion
  max-width: 100%;
```

  display: block;
}
```

#### Add the dimensions of the image to the img element {: #image-dimensions }

When using `max-width: 100%` you are overriding the natural dimensions of the image,
however you should still use the `width` and `height` attributes on your `<img>` tag.
This is because modern browsers will use this information to reserve space for the image
before it loads in,
this will help to avoid [layout shifts](/optimize-cls/) as content loads.

### Layout {: #layout }

Since screen dimensions and width in CSS pixels vary widely between devices
(for example, between phones and tablets, and even between different phones),
content should not rely on a particular viewport width to render well.

In the past, this required setting elements used to create layout in percentages.
In the example below, you can see a two-column layout with floated elements, sized using pixels.
Once the viewport becomes smaller than the total width of the columns, we have to scroll horizontally
to see the content.

<figure class="w-figure">

>   <img src="./fixed-width-layout.jpg" alt="Screenshot of a two-column layout with most of the second column outside the viewport">

Reviewer: ```suggestion
  <img class="w-screenshot" src="./fixed-width-layout.jpg" alt="Screenshot of a two-column layout with most of the second column outside the viewport">
```
* This will put a little border around the image and make it look better
* https://web.dev/handbook/markup-media/#partial-screenshots

  <figcaption class="w-figcaption">
    A floated layout using pixels. <a href="https://layout-floats-px.glitch.me/">See this example on Glitch</a>.
  </figcaption>
</figure>

By using percentages for the widths, the columns always remain a certain percentage of the container.
This means that the columns become narrower, rather than creating a scrollbar.

<div class="glitch-embed-wrap" style="height: 420px; width: 100%;">
  <iframe
    src="https://glitch.com/embed/#!/embed/layout-floats-percent?path=README.md&previewSize=100"
    title="layout-floats-percent on Glitch"
    style="height: 100%; width: 100%; border: 0;">
  </iframe>

> </div>

Reviewer: I added you to https://glitch.com/@webdev so you should be able to give that team access to the Glitch now

I'm getting a little cognitive dissonance with our recommendations here because the Glitch is overflowing the "viewport" (of the embed). How do you think we should address this?

![image](https://user-images.githubusercontent.com/4713486/82074216-4af0b480-968f-11ea-8999-07c32ee5f66d.png)

Author: I've added them to the team, still having issues getting the logo to stay on some of them. Sometimes it works and sometimes not.

The width issue was just my error, I'd remixed the fixed width example and neglected to change the width to max-width on the container. Should be sorted now.

}
```

In a responsive design, you can use Flexbox to display items as a single row,
or wrapped onto multiple rows as the available space decreases.

<div class="glitch-embed-wrap" style="height: 220px; width: 100%;">
  <iframe
    src="https://glitch.com/embed/#!/embed/responsive-flexbox?previewSize=100"
    title="responsive-flexbox on Glitch"
    style="height: 100%; width: 100%; border: 0;">
  </iframe>
</div>

[Read more about Flexbox](https://developer.mozilla.org/en-US/docs/Learn/CSS/CSS_layout/Flexbox).

#### CSS Grid Layout {: #grid }

CSS Grid Layout allows for the straightforward creation of flexible grids.
If we consider the earlier floated example,
rather than creating our columns with percentages,
we could use grid layout and the `fr` unit,
which represents a portion of the available space in the container.

```css
.container {
  display: grid;
  grid-template-columns: 1fr 3fr;
}
```

<div class="glitch-embed-wrap" style="height: 420px; width: 100%;">
  <iframe
    src="https://glitch.com/embed/#!/embed/two-column-grid?previewSize=100"
    title="two-column-grid on Glitch"
    style="height: 100%; width: 100%; border: 0;">
  </iframe>

> </div>

Reviewer: Same cognitive dissonance as the percentage float example:

![image](https://user-images.githubusercontent.com/4713486/82074342-7bd0e980-968f-11ea-8c97-2a82a7d53aa2.png)

Author: same issue, now fixed.

Forcing a user to switch to a mouse when they are using their touchscreen is not very friendly!
However, `any-hover` and `any-pointer` may be useful if it is important to work out what kind of device a user has.
For example, a laptop with a touchscreen and trackpad should match coarse and fine pointers,
in addition to the ability to hover.

## How to choose breakpoints {: #breakpoints }

Don't define breakpoints based on device classes.
Defining breakpoints based on specific devices, products, brand names,
or operating systems that are in use today can result in a maintenance nightmare.
Instead, the content itself should determine how the layout adjusts to its container.

### Pick major breakpoints by starting small, then working up {: #major-breakpoints }

Design the content to fit on a small screen size first,
then expand the screen until a breakpoint becomes necessary.
This allows you to optimize breakpoints based on content
and maintain the least number of breakpoints possible.

Let's work through the example we saw at the beginning: the weather forecast.
The first step is to make the forecast look good on a small screen.

<figure class="w-figure">

>   <img src="./weather-1-2x.png" width="400" alt="Screenshot of a weather app at a mobile width">

Reviewer: ```suggestion
  <img class="w-screenshot" src="./weather-1-2x.png" width="400" alt="Screenshot of a weather app at a mobile width">
```

  <figcaption class="w-figcaption">
    The app at a narrow width.
  </figcaption>
</figure>

Next, resize the browser until there is too much white space between the elements,
and the forecast simply doesn't look as good.
The decision is somewhat subjective, but above `600px` is certainly too wide.

<figure class="w-figure">

>   <img src="./weather-2-2x.png" width="400" alt="Screenshot of a weather app with wide gaps between items">

Reviewer: ```suggestion
  <img class="w-screenshot" src="./weather-2-2x.png" width="400" alt="Screenshot of a weather app with wide gaps between items">
```

Let's start by optimizing the small screen layout.
In this case, let's boost the font when the viewport width is greater than `360px`.
Second, when there is enough space,
we can separate the high and low temperatures so that they're on the same line
instead of on top of each other.
And let's also make the weather icons a bit larger.

```css
@media (min-width: 360px) {
  body {
    font-size: 1.0em;
  }
}

@media (min-width: 500px) {
  .seven-day-fc .temp-low,
  .seven-day-fc .temp-high {
    display: inline-block;
    width: 45%;
  }

  .seven-day-fc .seven-day-temp {
    margin-left: 5%;
  }

  .seven-day-fc .icon {
    width: 64px;
    height: 64px;
  }
}
```

Similarly, for the large screens it's best to limit to maximum width of the forecast panel
so it doesn't consume the whole screen width.

```css
@media (min-width: 700px) {
  .weather-forecast {
    width: 700px;
  }
}
```

<div class="glitch-embed-wrap" style="height: 420px; width: 100%;">
  <iframe
    src="https://glitch.com/embed/#!/embed/responsive-forecast?path=style.css&previewSize=100"
    title="responsive-forecast on Glitch"
    style="height: 100%; width: 100%; border: 0;">
  </iframe>

> </div>

Reviewer: So nice having the example embedded here. Thanks for migrating it.

### Optimize text for reading

Classic readability theory suggests that an ideal column should contain 70 to 80 characters per line
(about 8 to 10 words in English).
Thus, each time the width of a text block grows past about 10 words,
consider adding a breakpoint.

<figure class="w-figure">

>   <img src="./readability-narrow.png" width="400" alt="Screenshot of a a page of text on a mobile device">

Reviewer: ```suggestion
  <img class="w-screenshot" src="./readability-narrow.png" width="400" alt="Screenshot of a a page of text on a mobile device">
```

  <figcaption class="w-figcaption">
    The text as read on a mobile device.
  </figcaption>
</figure>

<figure class="w-figure">

>   <img src="./readability-wide.png" alt="Screenshot of a a page of text on a desktop browser">

Reviewer: ```suggestion
  <img class="w-screenshot" src="./readability-wide.png" alt="Screenshot of a a page of text on a desktop browser">
```

Let's take a deeper look at the above blog post example.
On smaller screens, the Roboto font at `1em` works perfectly giving 10 words per line,
but larger screens require a breakpoint.
In this case, if the browser width is greater than `575px`, the ideal content width is `550px`.

```css
@media (min-width: 575px) {
  article {
    width: 550px;
    margin-left: auto;
    margin-right: auto;
  }
}
```

<div class="glitch-embed-wrap" style="height: 420px; width: 100%;">
  <iframe
    src="https://glitch.com/embed/#!/embed/rwd-reading?path=index.html&previewSize=100"
    title="rwd-reading on Glitch"
    style="height: 100%; width: 100%; border: 0;">
  </iframe>
</div>

### Avoid simply hiding content

Be careful when choosing what content to hide or show depending on screen size.
Don't simply hide content just because you can't fit it on the screen.
Screen size is not a definitive indication of what a user may want.
For example, eliminating the pollen count from the weather forecast
could be a serious issue for spring-time allergy sufferers who need the information
to determine if they can go outside or not.

## View media query breakpoints in Chrome DevTools {: #devtools }

> Once you've got your media query breakpoints set up,
> you'll want to see how your site looks with them.
> You could resize your browser window to trigger the breakpoints,
> but there's a better way: Chrome DevTools.
> The two screenshots below demonstrate using DevTools to view how a page looks under different breakpoints.

Reviewer: ```suggestion
Once you've got your media query breakpoints set up,
you'll want to see how your site looks with them.
You could resize your browser window to trigger the breakpoints,
but Chrome DevTools has a built-in feature that makes it easy to see how a page looks
under different breakpoints.
```
I think I originally added this text to the guide and I'm now realizing that it sound sales-y

<figure class="w-figure">

>   <img src="./devtools-large.png" alt="Screenshot of DevTools with our weather app open and a width of 822 pixels selected.">

Reviewer: ```suggestion
  <img class="w-screenshot" src="./devtools-large.png" alt="Screenshot of DevTools with our weather app open and a width of 822 pixels selected.">
```

  <figcaption class="w-figcaption">
    DevTools showing the weather app as it looks at a wider viewport size.
  </figcaption>
</figure>

<figure class="w-figure">

>   <img src="./devtools-small.png" alt="Screenshot of DevTools with our weather app open and a width of 436 pixels selected.">

Reviewer: ```suggestion
  <img class="w-screenshot" src="./devtools-small.png" alt="Screenshot of DevTools with our weather app open and a width of 436 pixels selected.">
```

  <figcaption class="w-figcaption">
    An example of how the page loads in a device with the viewport meta tag. <a href="https://with-vp-meta.glitch.me/">See this example on Glitch</a>.
  </figcaption>
</figure>

[Some browsers](https://css-tricks.com/probably-use-initial-scale1/)
keep the page's width constant when rotating to landscape
mode, and zoom rather than reflow to fill the screen. Adding the value
`initial-scale=1` instructs browsers to establish a 1:1 relationship between CSS
pixels and device-independent pixels regardless of device orientation, and
allows the page to take advantage of the full landscape width.

{% Aside 'caution' %}
  To ensure that older browsers can properly parse the attributes,
  use a comma to separate attributes.
{% endAside %}

The [Does not have a `<meta name="viewport">` tag with `width` or `initial-scale`](/viewport/)
Lighthouse audit can help you automate the process of making sure that your HTML documents are using the viewport meta tag correctly.

### Ensure an accessible viewport {: #accessible-viewport }

In addition to setting an `initial-scale`,
you can also set the following attributes on the viewport:

> {% Compare 'worse' %}
>   * `minimum-scale`
>   * `maximum-scale`
>   * `user-scalable`
> {% endCompare %}

Reviewer: ```suggestion
* `minimum-scale`
* `maximum-scale`
* `user-scalable`
```

My suggestion isn't displaying correctly so it's probably easiest to just change it back to your original.

![image](https://user-images.githubusercontent.com/4713486/82338382-af6a8700-99a1-11ea-9300-8033c26177b0.png)
